nupic/research/TP.py

```diff
@@ -480,6 +480,335 @@ def diff(self, tp):
+
+    tp.version = proto.version
+    tp.verbosity = proto.verbosity
+    tp.consolePrinterVerbosity = proto.verbosity
+
+    tp._random = Random()
+    tp._random.read(proto.random)
+    tp.seed = proto.random.seed
+
+    tp.numberOfCols = proto.numberOfCols
+    tp.cellsPerColumn = proto.cellsPerColumn
+    tp._numberOfCells = proto.numberOfCols * proto.cellsPerColumn
+    tp.maxSegmentsPerCell = proto.maxSegmentsPerCell
+
+    tp.activationThreshold = proto.activationThreshold
+    tp.minThreshold = proto.minThreshold
+    tp.segUpdateValidDuration = proto.segUpdateValidDuration
+    tp.newSynapseCount = proto.newSynapseCount
+    tp.maxSynapsesPerSegment = proto.maxSynapsesPerSegment
+    tp.segID = 0
+
+    tp.initialPerm = numpy.float32(proto.initialPerm)
+    tp.connectedPerm = numpy.float32(proto.connectedPerm)
+    tp.permanenceMax = numpy.float32(proto.permanenceMax)
+    tp.permanenceDec = numpy.float32(proto.permanenceDec)
+    tp.permanenceInc = numpy.float32(proto.permanenceInc)
+
+    tp.doPooling = proto.doPooling
+    tp.globalDecay = numpy.float32(proto.globalDecay)
+    tp.maxAge = proto.maxAge
+    tp.maxInfBacktrack = proto.maxInfBacktrack
+    tp.maxLrnBacktrack = proto.maxLrnBacktrack
+    tp.pamLength = proto.pamLength
+    tp.pamCounter = proto.pamCounter
+    tp.learnedSeqLength = proto.learnedSeqLength
+    tp.avgLearnedSeqLength = proto.avgLearnedSeqLength
+    tp.maxSeqLength = proto.maxSeqLength
+    tp.avgInputDensity = proto.avgInputDensity
+    tp.outputType = proto.outputType
+    tp.burnIn = proto.burnIn
+    tp.collectStats = proto.collectStats
+    tp.collectSequenceStats = proto.collectSequenceStats
+
+    tp.iterationIdx = proto.iterationIdx
+    tp.lrnIterationIdx = proto.lrnIterationIdx
+    tp.resetCalled = proto.resetCalled
+    tp.segID = proto.nextSegIdx
+    tp.activeColumns = []
+
+    tp._initEphemerals()
+
+    lrnActiveStateT1Proto = proto.lrnActiveStateT1
+    lrnActiveStateTProto = proto.lrnActiveStateT
+    lrnPredictedStateT1Proto = proto.lrnPredictedStateT1
+    lrnPredictedStateTProto = proto.lrnPredictedStateT
+    infActiveStateT1Proto = proto.infActiveStateT1
+    infActiveStateTProto = proto.infActiveStateT
+    infActiveStateBackupProto = proto.infActiveStateBackup
+    infActiveStateCandidateProto = proto.infActiveStateCandidate
+    infPredictedStateT1Proto = proto.infPredictedStateT1
+    infPredictedStateTProto = proto.infPredictedStateT
+    cellConfidenceT1Proto = proto.cellConfidenceT1
+    cellConfidenceTProto = proto.cellConfidenceT
+    colConfidenceT1Proto = proto.colConfidenceT1
+    colConfidenceTProto = proto.colConfidenceT
+    cellsProto = proto.cells
+
+    tp.cells = []
+    for i in xrange(tp.numberOfCols):
+      tp.cells.append([])
+      for j in xrange(tp.cellsPerColumn):
+        tp.cells[i].append([])
+        cellAbsIdx = (i * tp.cellsPerColumn) + j
+        for segmentProto in cellsProto[cellAbsIdx].segments:
+          segment = object.__new__(Segment)
```

Comment:
Could you move this logic into the Segment class and use `Segment.read(segmentProto)`?